File: app/lib/models/ad_reward_session.dart
// 파일 의도: 서버가 발급하고 Google SSV가 완료하는 광고 보상 session 상태를 정의한다.

/// AdMob customData에 넣을 1회성 서버 token이다.
class AdRewardSession {
  const AdRewardSession({required this.sessionToken, required this.expiresAt});

  final String sessionToken;
  final DateTime expiresAt;

  factory AdRewardSession.fromJson(Map<String, Object?> json) {
    final token = json['sessionToken'];
    final expiresAt = json['expiresAt'];
    if (token is! String || token.isEmpty || expiresAt is! String) {
      throw const FormatException('Invalid ad reward session');
    }
    return AdRewardSession(
      sessionToken: token,
      expiresAt: DateTime.parse(expiresAt),
    );
  }
}

/// 광고 보상 session이 signed callback을 기다리는지, 완료됐는지, 만료됐는지를 나타낸다.
enum AdRewardStatus { pending, completed, expired }

/// 서버가 signed callback을 처리했는지 polling하는 상태다.
class AdRewardSessionStatus {
  const AdRewardSessionStatus({required this.status, required this.credited});

  final AdRewardStatus status;
  final bool? credited;

  factory AdRewardSessionStatus.fromJson(Map<String, Object?> json) {
    final rawStatus = json['status'];
    final credited = json['credited'];
    if (rawStatus is! String || (credited != null && credited is! bool)) {
      throw const FormatException('Invalid ad reward session status');
    }
    final status = switch (rawStatus) {
      'PENDING' => AdRewardStatus.pending,
      'COMPLETED' => AdRewardStatus.completed,
      'EXPIRED' => AdRewardStatus.expired,
      _ => throw const FormatException('Invalid ad reward status'),
    };
    return AdRewardSessionStatus(status: status, credited: credited as bool?);
  }
}

File: backend/src/main/java/com/lingko/lingko/core/domain/quota/entity/AdRewardSession.java
package com.lingko.lingko.core.domain.quota.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;

import java.time.Instant;

/** 앱 광고 한 건과 Google signed callback을 연결하는 1회성 서버 세션이다. */
@Entity
@Table(
        name = "ad_reward_sessions",
        uniqueConstraints = {
                @UniqueConstraint(columnNames = "session_token_hash"),
                @UniqueConstraint(columnNames = "transaction_id")
        }
)
@Getter
@Builder
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class AdRewardSession {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "ad_reward_session_id")
    private Long id;

    @Column(name = "user_idx", nullable = false)
    private Long userId;

    /** 원본 token은 앱에만 전달하고 DB에는 SHA-256만 저장한다. */
    @Column(name = "session_token_hash", nullable = false, length = 64)
    private String sessionTokenHash;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 16)
    private AdRewardSessionStatus status;

    /** Google이 서명한 전역 고유 지급 식별자다. */
    @Column(name = "transaction_id", length = 80)
    private String transactionId;

    @Column(name = "credited")
    private Boolean credited;

    @Column(name = "expires_at", nullable = false)
    private Instant expiresAt;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "completed_at")
    private Instant completedAt;

    public static AdRewardSession pending(Long userId, String tokenHash, Instant expiresAt) {
        if (userId == null || tokenHash == null || tokenHash.length() != 64 || expiresAt == null) {
            throw new IllegalArgumentException("ad reward session identity must be complete");
        }
        return AdRewardSession.builder()
                .userId(userId)
                .sessionTokenHash(tokenHash)
                .status(AdRewardSessionStatus.PENDING)
                .expiresAt(expiresAt)
                .build();
    }

    public boolean isExpiredAt(Instant now) {
        return status == AdRewardSessionStatus.PENDING && !now.isBefore(expiresAt);
    }

    public void expire() {
        if (status == AdRewardSessionStatus.PENDING) {
            status = AdRewardSessionStatus.EXPIRED;
        }
    }

    public void complete(String verifiedTransactionId, boolean wasCredited, Instant now) {
        if (status != AdRewardSessionStatus.PENDING) {
            return;
        }
        transactionId = verifiedTransactionId;
        credited = wasCredited;
        completedAt = now;
        status = AdRewardSessionStatus.COMPLETED;
    }
}
